### ISSUE TITLE
Ability to set canonical scale and level in MultiScaleRoIAlign

### ISSUE THREAD

**Issue by baldassarreFe** [pytorch/vision#3129]
## 🚀 Feature
A user should be able to set the parameters `canonical_scale` and `canonical_level` in the [torchvision.ops.poolers.LevelMapper](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L47) class when they instantiate a [torchvision.ops.poolers.MultiScaleRoIAlign](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L87) pooler.

## Motivation

The following is eq. 1 of Feature Pyramid Network [(page 4, bottom right)](https://arxiv.org/abs/1612.03144):
```
k = ⌊k0 + log2(√wh/224)⌋.
```
The authors write:
> Here `224` is the canonical ImageNet pre-training size, and `k0` is the target level on which an RoI with `w × h = 2242` should be mapped into. [...] we set k0 to 4.

In the code for multi-scale RoI align, these default values are reflected in the `canonical_scale` and `canonical_level` parameters of the [LevelMapper](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L47) class, which, given a box, decides which level of the pyramid to pool from.

The rationale behind these default values of 224 and 4 is related to the structure of a feature pyramid network. In fact, given an input image of `224x224`, a typical pyramid would look like this:
```python
{
  # C H W
  "input":  ( 3,  224, 224),
  "block1": (16,  112, 112),
  "block2": (32,   56,  56),
  "block3": (64,   28,  28),
  "block4": (128,  14,  14),
  "block5": (256,   7,   7),
}
```
With these sizes, it makes sense that a box with side 224 should be pooled from level 4 where it has been mapped into `14x14` patch.

However, when using a larger input size, one should be able to change these default values accordingly. For example, one could have an input image of `1280x800`, with the following pyramid:
```python
{
  # C H W
  "input":  ( 3, 800, 1280),
  "block1": (16, 400,  640),
  "block2": (32, 200,  320),
  "block3": (64, 100,  160),
  "block4": (128, 50,   80),
  "block5": (256, 25,   40),
  "block6": (256, 12,   20),
}
```
In this case, a box of size `224x224` could easily be pooled from the 6th block.

The ability to change this behavior, however, is currently hidden because when [MultiScaleRoIAlign](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L87) creates an instance of [LevelMapper](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L47) it sets the `k_min` and `k_max` parameters but leaves `canonical_scale` and `canonical_level` to their default values.

I was able to run a quick workaround by replacing the instance of [LevelMapper](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L47) after its creation. This workaround gives the expected results, i.e. I am able to decide which level to pool from depending on the area of the box and the size of the image.
```python
from math import sqrt
import torch
import torchvision as tv

height = 800
width = 1280
batch_size  = 2
fpn_features = {
    f"block{i}": torch.rand(batch_size, 2**(i+3), height // 2**i, width // 2**i)
    for i in [1, 2, 3]
}

roi_align = tv.ops.MultiScaleRoIAlign(
    featmap_names=list(fpn_features.keys()), 
    output_size=(3, 3), 
    sampling_ratio=2
)
roi_align.setup_scales(list(fpn_features.values()), image_shapes=[(height, width)] * batch_size)
roi_align.map_levels = tv.ops.poolers.initLevelMapper(
    k_min=roi_align.map_levels.k_min,
    k_max=roi_align.map_levels.k_max,
    canonical_scale=sqrt(width * height),
    canonical_level=4,
)
```

However, there should be an easier way for doing this.

## Pitch

To avoid modifying the code too much, I would suggest adding the  `canonical_scale` and `canonical_level` parameters to the constructor of [MultiScaleRoIAlign](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L87), which in turn will pass them to the `LevelMapper` [when it is created](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py#L179).

## Alternatives

The alternative solution would involve bigger changes to the logic in [`poolers.py`](https://github.com/pytorch/vision/blob/d3d393672b877f80fedd2d11de6b84fb19599c2e/torchvision/ops/poolers.py), possibly removing extra classes and initialization methods, and allowing the user more control. However, this clearly requires more discussion to get it right. One point that could be discussed is when it's best to compute the scales and initialize the mapping. Right now, they are initialized when `forward` is first called and persist in later calls, which might cause problems if the input sizes were to change. The issue is whether MultiScaleRoIAlign should behave more like a functional stateless operation (and hence allow changing image size at every call) or a stateful module (and hence compute fixed scales and mapping already in the constructor).

## Additional context

Here is a gist that demonstrates the issue: https://gist.github.com/baldassarreFe/513d798c089efa2c4feb0562dd1cfaf5

**Comment by fmassa** [comment]
Hi,

Thanks for such a detailed issue description, very comprehensive!

While I agree we haven't exposed `canonical_scale` and `canonical_level` to the user (and that it could be handy to expose them at some point), I'm not 100% sure I see their roles the same way you do (and I might be wrong here, so please correct me if I'm wrong).

From my understanding from the description in the paper, the `canonical_scale` and `canonical_level` should only be modified if we happened to change the underlying backbone to be trained on higher resolution images (for classification).
This is in some sense irrespective of the size of the image used for the detection purposes.

That being said, I think I would be ok exposing both arguments to the function, maybe making them a `kwargs`-only argument?

**Comment by baldassarreFe** [comment]
Thanks for the answer, I've had some more time to reason about the issue on my own. I realized that the use case for exposing those parameters is not exactly the one I described. Here are my thoughts:

### Scenario 1: typical backbone, higher res
We have a high-res backbone, let's say a typical ResNet that takes in images with a side of 2048 pixels. This backbone, despite the higher resolution, still follows the usual pattern of halving the output resolution at every block. In this case, if an object has a box of side 224 in the input, the corresponding patch will have shrunk to 224/2**4=14 by the fourth block, so it would make sense to pool its features from the fourth block

### Scenario 2: backbone with atypical downsizing
In this second scenario, the backbone follows an atypical downsizing pattern. For example, given an input with side 2048, a backbone that reduces the side by a factor of 4 at every block would produce:
- block1 512
- block2 128
- block3 32
- block4 8

In this case, the user might want to specify that an input patch with `canonical_scale=224` should be pooled from `canonical_level=2`, where it has already shrunk to 14.

---

With this said, I think it's still useful to expose those two parameters. As per your suggestion, I'll make them keyword-only and specify their meaning in the docstring.

In the future, a possible improvement would be to allow the user to instantiate a _mapper_ for their use case and pass it to the constructor. If no mapper is specified, we use the default one from the FPN paper, otherwise, any more complex behavior is put in the hands of the user.

**Comment by fmassa** [comment]
Merged in #3151
